src/Language/DifferentialDatalog/Validate.hs

```diff
@@ -90,34 +91,102 @@ checkNoRecursion :: (MonadError String me) => DatalogProgram -> me ()
 checkNoRecursion d = do
     case grCycle (funcGraph d) of
          Nothing -> return ()
-         Just t  -> err (pos $ getFunc d $ snd $ head t) 
+         Just t  -> err (pos $ getFunc d $ snd $ head t)
                         $ "Recursive function definition: " ++ (intercalate "->" $ map (name . snd) t)
 
 
 funcGraph :: DatalogProgram -> G.Gr String ()
-funcGraph DatalogProgram{..} = 
+funcGraph DatalogProgram{..} =
     let g0 = foldl' (\g (i,f) -> G.insNode (i,f) g)
                     G.empty $ zip [0..] (M.keys progFunctions) in
     foldl' (\g (i,f) -> case funcDef f of
                              Nothing -> g
-                             Just e  -> foldl' (\g' f' -> G.insEdge (i, M.findIndex f' progFunctions, ()) g') 
-                                               g (exprFuncs e)) 
+                             Just e  -> foldl' (\g' f' -> G.insEdge (i, M.findIndex f' progFunctions, ()) g')
+                                               g (exprFuncs e))
            g0 $ zip [0..] $ M.elems progFunctions
 
+-- convert FTL into rules
+convertFtl :: (MonadError String me) => DatalogProgram -> me DatalogProgram
+convertFtl d0 = do
+    rules <- mapM convertFtlStatement $ progStatements d0
+    return d0{progStatements = [], progRules = (progRules d0) ++ concat rules}
+
+-- given a list of match labels generates a matrix
+-- [a,b,c] -> [[(a,true), (b,false), (c,false)],
+--             [(a,false), (b,true), (c,false)],
+--             [(a,false), (b,false), (c,true)]]
+explodeMatchCases :: [Expr] -> [[(Expr, Expr)]]
+explodeMatchCases l =
+    let pairs = [ [(i1, i2) | i1 <- [0..(length l - 1)]] | i2 <- [0..(length l - 1)]] in
+    map (map (\(a,b) -> (l !! a, if a == b then eTrue else eFalse))) pairs
+
+-- add the specified RHS to all the rules
+addRhsToRules:: RuleRHS -> [Rule] -> [Rule]
+addRhsToRules toAdd rules =
+     map (\r -> Rule (rulePos r) (ruleLHS r) (toAdd : ruleRHS r)) rules
+
+convertFtlStatement :: (MonadError String me) => Statement -> me [Rule]
+convertFtlStatement (ForStatement p e r Nothing s) = do
+    rules <- convertFtlStatement s
+    let atom = Atom p r [("", e)]
+    let rhs = RHSLiteral True atom
+    let rules' = addRhsToRules rhs rules
+    return rules'
+convertFtlStatement (ForStatement p e r (Just c) s) = do
```

Comment:
One way to avoid two near-identical implementations for `ForStatement`:

```
convertFtlStatement (ForStatement p e r mc s) = do
    rules <- convertFtlStatement s
    let atom = Atom p r [("", e)]
    let rhs0 = RHSLiteral True atom
    let rhs1 = map RHSCondition $ maybeToList mc
    return $ map (\r -> r{ruleRHS = rhs0 : rhs1 ++ ruleRHS r}) rules
```